### DISCUSSION TITLE
Gswin64c.exe stays alive after Magick.Net thread is finished

### DISCUSSION THREAD

**Discussion by EL-bm** [dlemstra/Magick.NET#795] [Help] [answered]
Hello, hope everybody is doing great.
I have an app with the following features:
- Process multiple files using parallel programming.
   `Parallel.Foreach`
- Each thread opens up the file, creates an image collection, closes the file stream, and then continues.
   `var defines = new PdfReadDefines() { Interpolate = true };`
   `var settings = new MagickReadSettings();`
   `settings.Density = new Density(dpi);`
   `settings.Compression = CompressionMethod.NoCompression;`
   `settings.Format = MagickFormat.Pdf;`
   `settings.StrokeAntiAlias = true;`
   `settings.TextAntiAlias = true;`
   `settings.SetDefines(defines);`
   `using (images = new MagickImageCollection())`
   `{`
      `images.Read(_stream, settings);`
`_stream.Dispose();`
      `...`
   `}`
- When the thread process finishes/fails, the code makes sure the file stream gets properly closed and that the image collection gets cleared and set to null.
   `if (_stream != null)
            { 
               _stream.Dispose();
               _stream = null;
            }`
            `if (images != null)
            {
               images.Clear();
               images = null;
            }`
- Even with all the dispose mechanisms, the original files gets locked by a process gswin64c.exe (see image 1).

I know Magick.NET is capable of multi-threading and that only one thread will use the dll file and the rest will use the gswin64c.exe as command line.
The question: **Is there a way to make sure the Magick.NET instance also kills the gswin64c.exe process it created?**
Thanks in advance.

[Image 1]
![image](https://user-images.githubusercontent.com/71720005/100141565-42f0a900-2e58-11eb-9e35-ecd7651e8329.png)

**Comment by dlemstra** [comment]
Is it possible for you to create a small sample program (with a PDF) that can reproduce this consistently?

**Reply by EL-bm** [reply]
I am afraid we have not been able to find a pattern. 
Today we completed a set of tests (about 4 hours of continuous testing), always using the same 2 files and we got this problem twice. We don't know how to make it happen consistently.
We have seen our production servers to run out of space by the creation of temp files (we have seen over 15gbs consumed). The hard part is that there is no exception thrown, we realize when our integrity check systems starts telling us there is no space left.
I'll put a small sample code together as soon as possible.
Thanks for your fast response.

**Reply by EL-bm** [reply]
We just faced a massive application failure. 16 out of 20 servers went down with no warning, all of them with the same issue of no space and crowded with gswin64c.exe calls. Looking our logs we believe this is related to some of these errors:

- Error opening file 200-17389106 - PDFDelegateFailed `[ghostscript library 9.53] -q -dQUIET -dSAFER -dBATCH -dNOPAUSE -dNOPROMPT -dMaxBitmap=500000000 -dAlignToPixels=0 -dGridFitTT=2 "-sDEVICE=pngalpha" -dTextAlphaBits=4 -dGraphicsAlphaBits=4 "-r200x200"  "-sOutputFile=c:/tempFiles/magick-ab43qRweservh3I765ZUmN7iOBIdA7sg%d" "-fc:/tempFiles/magick-Xmo9Z7ooGNUq0JVknnzLYumBsjDRrVKx" "-fc:/tempFiles/magick-JUg3wPmZweYeMOGoJsh3tT1dW7uEPFRD": (null)' @ error/pdf.c/ReadPDFImage/660

- Error opening file 200-17393573 - FailedToExecuteCommand `"C:/bidmailapps/fps/ghostscript/gswin64c.exe" -q -dQUIET -dSAFER -dBATCH -dNOPAUSE -dNOPROMPT -dMaxBitmap=500000000 -dAlignToPixels=0 -dGridFitTT=2 "-sDEVICE=pngalpha" -dTextAlphaBits=4 -dGraphicsAlphaBits=4 "-r200x200"  "-sOutputFile=c:/tempFiles/magick-w5UDhtPRul46javNjTP9CgSHAJbzzl2V%d" "-fc:/tempFiles/magick-olu6iC0vzKHPB8pELgn4UhRCi9id8Nsc" "-fc:/tempFiles/magick-HjLoq5sTCtdpRj-XfC6ZkJY7hK6G-Mqr"' (1) @ error/ghostscript-private.h/InvokeGhostscriptDelegate/150

It looks like if one of these errors happens the gswin64c.exe will be kept alive.

**Reply by EL-bm** [reply]
After a loooong day of having issues with this topic, we finally got to a point where every file being processed with the gswin64c.exe was simply failing silently (the process will just idle blocking other resources). We went to ghostscript site and we found a new version (we are using 9.53.1 -> 9.53.3) and we tried in our alpha servers with no difference.
The only reliable method is to have the files processed with the dll so there it goes our fantastic multi-threading approach :(
Now I am wondering **if it is possible to specify the ghostscript folder for each MagickImageCollection.Read so each read operation can use a different copy (on different location) of the ghostscript dll?**
I know I have written a lot of comments today but any thoughts are more than welcome and greatly appreciated.
Thanks in advance.

**Reply by dlemstra** [reply]
You cannot choose a different folder for each execution. Have you checked if the files in the tempFiles folder are properly cleaned up? Maybe the process hangs because you are running out of disk space? It might be possible that those two problem images are super large and forcing MagickNET to swap the pixels to disk. You probably want to limit the max width and height by configuring that in a policy or setting `ResourceLimits.Width` and `Height` in runtime.

**Reply by EL-bm** [reply]
Thanks @dlemstra. Our solution for now is simply avoid using the gswin64c.exe and the way we were able to achieve that is to create copies of the same service and each copy uses a new location of the dll. So we have 4 services per server ,TileCreator1 - 4, and 4 different copies of the dll component so each instance of the services calls a different one. Each server process one file a time so we process 4 files per server (kind of multi-threading but in the hard way).

To answer your questions:
1.- Are we running out of space? -> Yes but the reason is the gswin64c.exe process not dying. Since the exe process persists forever (until the service gets rebooted altogether), the temp files are locked forever. We have a special sub-process in charge of cleaning any temp files that are not being used and that are older than 10 mins.

My apologies to the team since I have had not found the time to write a sample code illustrating this issue mainly because I was forced to look for a quick solution since this is our production process. 

- I'll try to find some time to write a sample app. 
- Until then it is my opinion there is a problem in the multithreading mechanism of the Magick.NET library **BECAUSE** it is using the .exe. You guys mentioned in other post that it is not possible to use the dll for all of the threads and I don't really know why.
- Given our results using four different copies of the dll, I strongly believe that this kind of issues would be resolved if Magick.Net could internally use the dll for all the threads.

I don't know if you guys want to mark this thread as closed. Let me know.

**Reply by dlemstra** [reply]
Having 4 different processes is different from having 4 threads in the same process. The library is not thread safe and that is why the next execution will need to use the binary instead of the library.

Is it possible that the file that causes this issues is corrupt and causes `gswin64c.exe` to never exit?

**Reply by EL-bm** [reply]
File corruption is what we review in second place being pdf world a "diverse" one. We looked for incomplete fonts, different types of images and metadata and even custom signatures for bookmarks. What we found is that there is no apparent pattern based on the pdf structure which means the process could fail for a file that would be later process correctly on a second try. 
Here is a list of what we checked:
- Antivirus
- Antiramsomware
- PDF corruption
- Network latency and disconnection (our files are located on several network locations)
- Interference by the edition department (a separate group of people works over these same pdf files)
- Interference by our main viewer app (the one that uses the created tiles)
- Tile creator service logic
- Tile creator service interaction with core db (looking for deadlocks that would prevent the proper update of the data causing malfunctioning)
- Magick.NET library

We turned our heads to the Magick.NET library at the end because it showed no issues during development phase. We have been using the Magick.NET library since 2 years ago for creating the tiles but at that time we were using a different component to read a raster version of a PDF. It was until 2 months ago that we started using Magick.Net library combined with ghostscript to also read the raster from the pdf.

After this extensive review we were able to determine that the only common element for most of the issues was the use of the gswin64c.exe. Using the exe will fail for the same file that the dll (or even a different thread using the exe) will process correctly. We don't know why there is such disparity when using the exe not being really experts on ghostscript command line.
We process a volume around 2k - 4k pdf files a day. For our testing we focused on the files we knew were a fail getting a sample space of 44 files. At the end, we were able to replicate the problem using even just 2 files (2 needed to trigger the usage of the binary).

The following questions float on my mind:
- Why the dll is not failing (so far using the dll is not showing issues) and the binary is? This looks like a problem on ghostscript side.
- Why is the binary failing? Even for the same file it'll fail some times and some others won't. Ghostscript issue maybe?
- When the binary is failing, why is the binary execution not getting terminated? This doesn't happen all the time but it is the core issue for us since the original file is getting blocked (preventing other instances to process it) and the temp files are also getting locked eating up the hard drive space eventually leading to the server dying. This looks like on Magick.NET issue.
 
I am not allowed to share a direct version of our code (NDA) so I am looking for some time to write a code sample so you guys can analyze it but it is a bit hectic being the end of the year and all.

**Reply by EL-bm** [reply]
Dear @dlemstra ,
Here is a link to download the sample code:
https://www.dropbox.com/s/mrt1roo32wi70pk/TileCreatorSample.zip?dl=0
Sorry for the delay. Please unpack the file and read the instructions.
Hope this helps.

**Comment by EL-bm** [comment]
In case nobody saw my previous comment from 4 days ago:
Dear @dlemstra ,
Here is a link to download the sample code:
https://www.dropbox.com/s/mrt1roo32wi70pk/TileCreatorSample.zip?dl=0
Sorry for the delay. Please unpack the file and read the instructions.
Hope this helps.

**Reply by dlemstra** [reply]
Seen your message, haven't had/made time for this yet.

**Reply by dlemstra** [reply]
I finally made some time to investigate this. It turns out that the process is in a deadlock. We are waiting for the process to exit but the process is waiting for us to read it's output. One of your test files spits out a lot of messages to the output and if I force that file to always use the command line a deadlock will occur. Thanks for providing this example and a fix will become available in the next version of Magick.NET

**Reply by EL-bm** [reply]
No better way to start 2021 than contribute with others in someway!! I am happy you guys were able to spot this one. I believe it was a tricky error since specific conditions are needed for it to reveal. Let me know if I can be of any assistance. 
I'll wait for the next iteration of this awesome library.
You guys have a happy and blessed 2021.
Best regards!

**Comment by tomasr78** [comment]
I have exactly the same problem - sometimes the produced Image files from PDF are being locked and released after one or two seconds or even never released.  I have tried to update to the new `Magick.NET 7.22.3.0` version but still, the problem persists. I am using this code to execute pdf to image in parallel https://github.com/dlemstra/Magick.NET/discussions/810

I do not find a pattern of how to reproduce the problem, it just happens without a clear reason.

**Comment by dlemstra** [comment]
Your description does not sound like the exact same problem. PDF files being locked is different from the Ghostscript executable not exiting.

p.s. @EL-bm  Can you confirm that this issue is resolved?

**Reply by tomasr78** [reply]
Ups, sorry, yes the problem may be different and not PDF but created PNG is being locked. I am still trying to debug and find more info.

**Reply by EL-bm** [reply]
Woops, sorry, been so busy lately that I did not notice there was a new version. I'll try to implement this new version asap. I'll let you guys know.
Thanks for all your hard work.

**Reply by EL-bm** [reply]
Dear @dlemstra and @tomasr78,
Today we installed the new version in our alpha servers. Everything seems to be working fine. I see the exe app getting disposed off correctly and our app is now running in a single service instance with multithreading.
For a better details, we upgraded our Magick.NET version to the current one: 7.23.0.0.
We'll continue monitoring this new version for a week and a half. I'll report back on Friday (or earlier if any problem) with more details.
Hope this info helps.

**Comment by EL-bm** [comment, marked as answer]
New update,
The service using the new Magick.Net version (7.23.0.0)  got heavily tested yesterday. Thanks to another requirement we got the chance the process a volume over 500k files and I am happy to report that we did not see any of the previous issues.
Being all this said, this bug is fixed in my opinion.
Thank you all for your hard work, this library is amazing in many ways.

**Comment by dlemstra** [comment]
Happy to hear this got resolved and thank you for helping me fix this.
